pkg/localstore/subscription_push.go

```diff
@@ -105,6 +120,23 @@ func (db *DB) SubscribePush(ctx context.Context) (c <-chan swarm.Chunk, stop fun
 					db.logger.Debugf("localstore push subscription iteration: %v", err)
 					return
 				}
+
+				// save last Item from this iteration in order to know were
+				// to stop on next iteration
```

Comment:
i dont really understand the comment. the last item from this iteration is actually where we _start_ the next iteration no?

Reply:
Actually no, last item is temporary variable, used to save first received item from iterator, which should be currently last one in the database.
As far as I could find through the code, once started, all elements are being processed from `pushIndex` once started.
If all are processed, first one that was received (which is last in the database), is copied to `toItemKey` variable, which is used to compare on next loops for iteration to know on which item was already processed before (depending on the reason why the iteration stopped).

In the example where we start with this initial elements in the database:

```txt
01
02
03
```

first item processed will be `03`, and will be stored in `lastItem` variable. Next item will be `02` and then `01`.
If some new items are added in the meantime, we might come to situation like this (assuming no removal of previous items happened):

```txt
01
02
03
04
05
```

On the next iteration first item will be `05`, and we again go backward, to `04`. However once we come to `03` we do not want to process that item again, as we already did. That is why each item key is compared to `toItemKey`, and why we stop iteration in that case.

This should basically result in same process that was here previously, where iteration was being started from last processed chunk.

We cannot configure option to start from some other item, as that would mean starting from some middle of items. We always want to start from last added item.
I have considered having that option in addition to these changes, however I could not find condition where this would actually be used, which is why it was removed.

Reply:
In theory it is possible. However, as I noted in the comment, I have gone through all the cases that I could, and could not find scenario where that happens. If any error occurs, it should complete enclosing goroutine.

As far as I understood previous code, it basically iterates through all items, until it reaches end (processes all elements), and then when triggered starts from last processed item (previously used `sinceItem` variable). If there was some way that iteration could stop prematurely and without an error I would like to know, to make sure I do no miss it. As for now, I could not find it.

New code should to the same, only in reversed order. Once iteration starts, it should go through all elements until it reaches end (sorted first key in LevelDB). On the next trigger, it should start from last element in the index, and at most to the previous element that was already processed (skipping it).

Only possible scenario that I could imagine is that somehow during iteration of LevelDB, some element in the middle of the queue is removed (before it is processed), and `leveldb.ErrNotFound` error is reached, at which point internal iteration loop is stopped, at which point we have a gap in the middle, which should require re-subscription to start whole process again (which does happen in some cases).
However I could not find case when random items are removed from push index before being processed.

Reply:
cool. very thorough :+1: thanks